Diabeties Dataset Cleaning and Exploration

| Feature Name               | Type    | Description | % Missing |
|----------------------------|---------|-------------|-----------|
| Encounter ID               | Numeric | Unique identifier of an encounter | 0% |
| Patient number             | Numeric | Unique identifier of a patient | 0% |
| Race                       | Nominal | Values: Caucasian, Asian, African American, Hispanic, and other | 2% |
| Gender                     | Nominal | Values: male, female, and unknown/invalid | 0% |
| Age                        | Nominal | Grouped in 10-year intervals: [0, 10), [10, 20), ..., [90, 100) | 0% |
| Weight                     | Numeric | Weight in pounds | 97% |
| Admission type             | Nominal | Integer identifier corresponding to 9 distinct values, e.g., emergency, urgent, elective, newborn, and not available | 0% |
| Discharge disposition      | Nominal | Integer identifier corresponding to 29 distinct values, e.g., discharged to home, expired, and not available | 0% |
| Admission source           | Nominal | Integer identifier corresponding to 21 distinct values, e.g., physician referral, emergency room, and transfer from a hospital | 0% |
| Time in hospital           | Numeric | Integer number of days between admission and discharge | 0% |
| Payer code                 | Nominal | Integer identifier corresponding to 23 distinct values, e.g., Blue Cross/Blue Shield, Medicare, and self-pay | 52% |
| Medical specialty          | Nominal | Integer identifier of the admitting physician's specialty (84 distinct values), e.g., cardiology, internal medicine, family/general practice, and surgeon | 53% |
| Number of lab procedures   | Numeric | Number of lab tests performed during the encounter | 0% |
| Number of procedures       | Numeric | Number of procedures (other than lab tests) performed during the encounter | 0% |
| Number of medications      | Numeric | Number of distinct generic names administered during the encounter | 0% |
| Number of outpatient visits | Numeric | Number of outpatient visits of the patient in the year preceding the encounter | 0% |
| Number of emergency visits | Numeric | Number of emergency visits of the patient in the year preceding the encounter | 0% |
| Number of inpatient visits | Numeric | Number of inpatient visits of the patient in the year preceding the encounter | 0% |
| Diagnosis 1               | Nominal | Primary diagnosis (coded as first three digits of ICD9); 848 distinct values | 0% |
| Diagnosis 2               | Nominal | Secondary diagnosis (coded as first three digits of ICD9); 923 distinct values | 0% |
| Diagnosis 3               | Nominal | Additional secondary diagnosis (coded as first three digits of ICD9); 954 distinct values | 1% |
| Number of diagnoses       | Numeric | Number of diagnoses entered to the system | 0% |
| Glucose serum test result | Nominal | Indicates the range of the result or if the test was not taken. Values: ">200", ">300", "normal", "none" | 0% |
| A1c test result           | Nominal | Indicates the range of the result or if the test was not taken. Values: ">8", ">7", "normal", "none" | 0% |
| Change of medications     | Nominal | Indicates if there was a change in diabetic medications. Values: "change" and "no change" | 0% |
| Diabetes medications      | Nominal | Indicates if any diabetic medication was prescribed. Values: "yes" and "no" | 0% |
| 24 features for medications | Nominal | Each feature represents a drug (e.g., metformin, insulin) and indicates whether it was prescribed or dosage changed. Values: "up", "down", "steady", "no" | 0% |
| Readmitted                | Nominal | Days to inpatient readmission. Values: "<30", ">30", "No" | 0% |


In [41]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [77]:
data = pd.read_csv('diabetic_data.csv')

In [79]:
data.columns

Index(['encounter_id', 'patient_nbr', 'race', 'gender', 'age', 'weight',
       'admission_type_id', 'discharge_disposition_id', 'admission_source_id',
       'time_in_hospital', 'payer_code', 'medical_specialty',
       'num_lab_procedures', 'num_procedures', 'num_medications',
       'number_outpatient', 'number_emergency', 'number_inpatient', 'diag_1',
       'diag_2', 'diag_3', 'number_diagnoses', 'max_glu_serum', 'A1Cresult',
       'metformin', 'repaglinide', 'nateglinide', 'chlorpropamide',
       'glimepiride', 'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide',
       'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone',
       'tolazamide', 'examide', 'citoglipton', 'insulin',
       'glyburide-metformin', 'glipizide-metformin',
       'glimepiride-pioglitazone', 'metformin-rosiglitazone',
       'metformin-pioglitazone', 'change', 'diabetesMed', 'readmitted'],
      dtype='object')

In [ ]:
[data['diag_1'] > 250]

array(['250.83', '276', '648', '8', '197', '414', '428', '398', '434',
       '250.7', '157', '518', '999', '410', '682', '402', '737', '572',
       'V57', '189', '786', '427', '996', '277', '584', '462', '473',
       '411', '174', '486', '998', '511', '432', '626', '295', '196',
       '250.6', '618', '182', '845', '423', '808', '250.4', '722', '403',
       '250.11', '784', '707', '440', '151', '715', '997', '198', '564',
       '812', '38', '590', '556', '578', '250.32', '433', 'V58', '569',
       '185', '536', '255', '250.13', '599', '558', '574', '491', '560',
       '244', '250.03', '577', '730', '188', '824', '250.8', '332', '562',
       '291', '296', '510', '401', '263', '438', '70', '250.02', '493',
       '642', '625', '571', '738', '593', '250.42', '807', '456', '446',
       '575', '250.41', '820', '515', '780', '250.22', '995', '235',
       '250.82', '721', '787', '162', '724', '282', '514', 'V55', '281',
       '250.33', '530', '466', '435', '250.12', 'V53', '789', '

In [ ]:
data[data['Diagnosis1'] == ''].shape

(11357, 50)

In [44]:
data.columns

Index(['encounter_id', 'patient_nbr', 'race', 'gender', 'age', 'weight',
       'admission_type_id', 'discharge_disposition_id', 'admission_source_id',
       'time_in_hospital', 'payer_code', 'medical_specialty',
       'num_lab_procedures', 'num_procedures', 'num_medications',
       'number_outpatient', 'number_emergency', 'number_inpatient', 'diag_1',
       'diag_2', 'diag_3', 'number_diagnoses', 'max_glu_serum', 'A1Cresult',
       'metformin', 'repaglinide', 'nateglinide', 'chlorpropamide',
       'glimepiride', 'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide',
       'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone',
       'tolazamide', 'examide', 'citoglipton', 'insulin',
       'glyburide-metformin', 'glipizide-metformin',
       'glimepiride-pioglitazone', 'metformin-rosiglitazone',
       'metformin-pioglitazone', 'change', 'diabetesMed', 'readmitted'],
      dtype='object')

In [45]:
data.shape

(101766, 50)

In [46]:
data['weight'].unique()

array(['?', '[75-100)', '[50-75)', '[0-25)', '[100-125)', '[25-50)',
       '[125-150)', '[175-200)', '[150-175)', '>200'], dtype=object)

In [47]:
data.shape

(101766, 50)

In [48]:
data[data['weight'] == '?']

encounter_id  patient_nbr             race  gender      age weight  \
0            2278392      8222157        Caucasian  Female   [0-10)      ?   
1             149190     55629189        Caucasian  Female  [10-20)      ?   
2              64410     86047875  AfricanAmerican  Female  [20-30)      ?   
3             500364     82442376        Caucasian    Male  [30-40)      ?   
4              16680     42519267        Caucasian    Male  [40-50)      ?   
...              ...          ...              ...     ...      ...    ...   
101761     443847548    100162476  AfricanAmerican    Male  [70-80)      ?   
101762     443847782     74694222  AfricanAmerican  Female  [80-90)      ?   
101763     443854148     41088789        Caucasian    Male  [70-80)      ?   
101764     443857166     31693671        Caucasian  Female  [80-90)      ?   
101765     443867222    175429310        Caucasian    Male  [70-80)      ?   

        admission_type_id  discharge_disposition_id  admission_source_id  \
0                       6                        25                    1   
1                       1                         1                    7   
2                       1                         1                    7   
3                       1                         1                    7   
4                       1                         1                    7   
...                   ...                       ...                  ...   
101761                  1                         3                    7   
101762                  1                         4                    5   
101763                  1                         1                    7   
101764                  2                         3                    7   
101765                  1                         1                    7   

        time_in_hospital  ... citoglipton insulin  glyburide-metformin  \
0                      1  ...          No      No                   No   
1                      3  ...          No      Up                   No   
2                      2  ...          No      No                   No   
3                      2  ...          No      Up                   No   
4                      1  ...          No  Steady                   No   
...                  ...  ...         ...     ...                  ...   
101761                 3  ...          No    Down                   No   
101762                 5  ...          No  Steady                   No   
101763                 1  ...          No    Down                   No   
101764                10  ...          No      Up                   No   
101765                 6  ...          No      No                   No   

        glipizide-metformin  glimepiride-pioglitazone  \
0                        No                        No   
1                        No                        No   
2                        No                        No   
3                        No                        No   
4                        No                        No   
...                     ...                       ...   
101761                   No                        No   
101762                   No                        No   
101763                   No                        No   
101764                   No                        No   
101765                   No                        No   

        metformin-rosiglitazone  metformin-pioglitazone  change diabetesMed  \
0                            No                      No      No          No   
1                            No                      No      Ch         Yes   
2                            No                      No      No         Yes   
3                            No                      No      Ch         Yes   
4                            No                      No      Ch         Yes   
...                         ...                     ...     ...         ...   
101761                       No

98569 weights are missing out of 101766. Do not use weight as a factor for prediction model

In [49]:
data.isna().sum().sort_values(ascending=False)

encounter_id                0
troglitazone                0
chlorpropamide              0
glimepiride                 0
acetohexamide               0
glipizide                   0
glyburide                   0
tolbutamide                 0
pioglitazone                0
rosiglitazone               0
acarbose                    0
miglitol                    0
tolazamide                  0
patient_nbr                 0
examide                     0
citoglipton                 0
insulin                     0
glyburide-metformin         0
glipizide-metformin         0
glimepiride-pioglitazone    0
metformin-rosiglitazone     0
metformin-pioglitazone      0
change                      0
diabetesMed                 0
nateglinide                 0
repaglinide                 0
metformin                   0
A1Cresult                   0
race                        0
gender                      0
age                         0
weight                      0
admission_type_id           0
discharge_

max_glu_serum and a1cresult should not be used either, lots of missing values

Filter dataset to only include diabetes diagnoses

In [50]:
def is_numeric(s):
    try:
        float(s)
        return True
    except ValueError:
        return False

In [51]:
data['diabetes_1'] = data['diag_1'].apply(lambda x: float(x) if is_numeric(x) else np.nan)
data['diabetes_2'] = data['diag_2'].apply(lambda x: float(x) if is_numeric(x) else np.nan)
data['diabetes_3'] = data['diag_3'].apply(lambda x: float(x) if is_numeric(x) else np.nan)

In [52]:
data = data[((data['diabetes_1'] > 250) & (data['diabetes_1'] < 251)) 
          | ((data['diabetes_2'] > 250) & (data['diabetes_2'] < 251)) 
          | ((data['diabetes_3'] > 250) & (data['diabetes_3'] < 251))]


Replace weight '?' with np.nan

In [53]:
data['weight'].replace('?', np.nan, inplace=True)

In [54]:
data

encounter_id  patient_nbr             race  gender      age weight  \
0            2278392      8222157        Caucasian  Female   [0-10)    NaN   
1             149190     55629189        Caucasian  Female  [10-20)    NaN   
3             500364     82442376        Caucasian    Male  [30-40)    NaN   
10             28236     89869032  AfricanAmerican  Female  [40-50)    NaN   
12             40926     85504905        Caucasian  Female  [40-50)    NaN   
...              ...          ...              ...     ...      ...    ...   
101736     443739152     90751788        Caucasian  Female  [60-70)    NaN   
101742     443787512     52419276        Caucasian    Male  [70-80)    NaN   
101753     443841992    184875899            Other    Male  [40-50)    NaN   
101754     443842016    183087545        Caucasian  Female  [70-80)    NaN   
101761     443847548    100162476  AfricanAmerican    Male  [70-80)    NaN   

        admission_type_id  discharge_disposition_id  admission_source_id  \
0                       6                        25                    1   
1                       1                         1                    7   
3                       1                         1                    7   
10                      1                         1                    7   
12                      1                         3                    7   
...                   ...                       ...                  ...   
101736                  1                         3                    7   
101742                  2                         6                    2   
101753                  1                         1                    7   
101754                  1                         1                    7   
101761                  1                         3                    7   

        time_in_hospital  ... glipizide-metformin glimepiride-pioglitazone  \
0                      1  ...                  No                       No   
1                      3  ...                  No                       No   
3                      2  ...                  No                       No   
10                     9  ...                  No                       No   
12                     7  ...                  No                       No   
...                  ...  ...                 ...                      ...   
101736                 8  ...                  No                       No   
101742                 4  ...                  No                       No   
101753                13  ...                  No                       No   
101754                 9  ...                  No                       No   
101761                 3  ...                  No                       No   

        metformin-rosiglitazone  metformin-pioglitazone  change  diabetesMed  \
0                            No                      No      No           No   
1                            No                      No      Ch          Yes   
3                            No                      No      Ch          Yes   
10                           No                      No      No          Yes   
12                           No                      No      Ch          Yes   
...                         ...                     ...     ...          ...   
101736                       No                      No      No          Yes   
101742                       No                      No      Ch          Yes   
101753                       No                      No      Ch          Yes   
101754                       No                      No      Ch          Yes   
101761                       No                      No      Ch          Yes   

        readmitted  diabetes_1 diabetes_2 diabetes_3  
0               NO      250.83        NaN        NaN  
1              >30      276.00     250.01     255.00  
3               NO        8.00     250.43     403.00  
10             >30      250.70     403.0

Handling missing values

In [55]:
data.isna().sum().sort_values(ascending=False)

weight                      19798
diabetes_3                   1621
diabetes_2                    636
diabetes_1                    195
examide                         0
acetohexamide                   0
glipizide                       0
glyburide                       0
tolbutamide                     0
pioglitazone                    0
rosiglitazone                   0
acarbose                        0
miglitol                        0
troglitazone                    0
tolazamide                      0
insulin                         0
citoglipton                     0
chlorpropamide                  0
glyburide-metformin             0
glipizide-metformin             0
glimepiride-pioglitazone        0
metformin-rosiglitazone         0
metformin-pioglitazone          0
change                          0
diabetesMed                     0
readmitted                      0
glimepiride                     0
encounter_id                    0
patient_nbr                     0
num_procedures

We leave diabeties 1, 2, 3 na values as is because we should not impute a diagnoses. We are only trying to focus on the diagnoses of diabeties

Look for '?' and replace with na values

In [56]:
data.replace('?', np.nan, inplace=True)

Remove irrelevant columns

'Encounter_id': We will use index as unique identifier

'Diagnosis 1,2,3': We filtered our dataset to if a paitent was diagnosed with diabeties either or through 1,2,3

'Weights' has to many missing values to be useful so, it can be removed.

'Payer_code' has half missing values so it can be removed.

'Medical_specialty' has many missing values so it can also be removed.

In [57]:
missing_data = []
for column in data.columns:
    if data[column].isna().sum() / data.shape[0] > 0.1:
        missing_data.append(column)
missing_data

['weight', 'payer_code', 'medical_specialty']

In [58]:
data = data.drop(['encounter_id', 'diag_1', 'diag_2', 'diag_3','weight','payer_code','medical_specialty'], axis=1)

Make sure data types are correct

Age and weight are binned, so strings instead of ints

Diabeties 1, 2, 3 are IC9 codes which all mean this patient was diagnosed with diabetes

Medication columns include values 'Yes', 'No', 'Steady', 'Up', 'Down'. We did not convert these columns to 1,0 because there are multiple values, not just yes, no

In [59]:
data.columns 

Index(['patient_nbr', 'race', 'gender', 'age', 'admission_type_id',
       'discharge_disposition_id', 'admission_source_id', 'time_in_hospital',
       'num_lab_procedures', 'num_procedures', 'num_medications',
       'number_outpatient', 'number_emergency', 'number_inpatient',
       'number_diagnoses', 'max_glu_serum', 'A1Cresult', 'metformin',
       'repaglinide', 'nateglinide', 'chlorpropamide', 'glimepiride',
       'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide',
       'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone',
       'tolazamide', 'examide', 'citoglipton', 'insulin',
       'glyburide-metformin', 'glipizide-metformin',
       'glimepiride-pioglitazone', 'metformin-rosiglitazone',
       'metformin-pioglitazone', 'change', 'diabetesMed', 'readmitted',
       'diabetes_1', 'diabetes_2', 'diabetes_3'],
      dtype='object')

One hot encode all categorical , drop the medicines which are within the varse minority of the population.

In [60]:
data = data.drop(['acetohexamide','tolbutamide','miglitol','chlorpropamide','acarbose','miglitol','troglitazone','tolazamide','examide','citoglipton','glyburide-metformin','glipizide-metformin','glimepiride-pioglitazone','metformin-rosiglitazone','metformin-pioglitazone'], axis=1)

In [61]:
def one_hot_encode(df, column):
    one_hot = pd.get_dummies(df[column], prefix=column)
    df = df.drop(column, axis=1)
    df = df.join(one_hot)
    return df

In [62]:
data['change'] = data['change'].replace('Ch', 1).replace('No', 0)
data['diabetesMed'] = data['diabetesMed'].replace('Yes', 1).replace('No', 0)

In [63]:
to_one_hot_encode = ['race','gender','age','max_glu_serum','A1Cresult','metformin','repaglinide','nateglinide','glimepiride','glipizide','glyburide','pioglitazone','rosiglitazone','insulin']

In [64]:
data['readmitted'] = data['readmitted'].replace({'NO': 0, '<30': 1, '>30': 1})

In [65]:
one_hot_encoded_data = data.copy()
for i in to_one_hot_encode:
    one_hot_encoded_data = one_hot_encode(one_hot_encoded_data, i)

Final Dataset

In [66]:
one_hot_encoded_data.shape

(20166, 79)

In [67]:
def plot_corr(data):
    f = plt.figure(figsize=(19,10))
    plt.matshow(data.corr(), fignum=f.number, cmap='RdBu')
    plt.xticks(range(data.select_dtypes(['number']).shape[1]), data.select_dtypes(['number']).columns, fontsize=14, rotation=45)
    plt.yticks(range(data.select_dtypes(['number']).shape[1]), data.select_dtypes(['number']).columns, fontsize=14)
    cb = plt.colorbar()
    cb.ax.tick_params(labelsize=14)
    plt.title('Correlation Matrix', fontsize=16)

In [68]:
one_hot_encoded_data.corr()['readmitted'].sort_values(ascending=False)[1:23]

number_inpatient       0.232918
number_diagnoses       0.137896
number_emergency       0.120588
patient_nbr            0.099129
number_outpatient      0.094769
num_medications        0.088336
A1Cresult_None         0.073984
time_in_hospital       0.064675
metformin_No           0.060842
race_Caucasian         0.044467
admission_source_id    0.039107
insulin_Down           0.034138
diabetes_2             0.031736
age_[70-80)            0.031370
admission_type_id      0.029890
gender_Female          0.029847
max_glu_serum_>300     0.028263
glyburide_No           0.022961
insulin_Up             0.018671
max_glu_serum_>200     0.018635
age_[60-70)            0.016485
max_glu_serum_Norm     0.016145
Name: readmitted, dtype: float64

#### Check for Multicollinearity

In [69]:
from statsmodels.stats.outliers_influence import variance_inflation_factor

In [70]:
def calculate_vid(X):
    # calculating VIF 
    vif = pd.DataFrame()
    vif["variables"] = X.columns
    vif["VIF"] = [variance_inflation_factor(X.values, i) for i in range(X.shape[1])]
    return vif

In [71]:
all_cols = one_hot_encoded_data.isna().sum(axis = 0) != 0

# columns that still contain null value
null_cols = all_cols[all_cols == True].index 

In [72]:
data_copy = one_hot_encoded_data.copy().drop(columns = null_cols)
data_vif = calculate_vid(data_copy)

c:\Users\r2d2d\miniforge3\envs\dsc80\lib\site-packages\statsmodels\stats\outliers_influence.py:195: RuntimeWarning: divide by zero encountered in double_scalars
  vif = 1. / (1. - r_squared_i)


In [73]:
data_vif.sort_values("VIF")[:20]

variables        VIF
8          number_outpatient   1.057049
14                readmitted   1.099402
2   discharge_disposition_id   1.114011
9           number_emergency   1.171147
0                patient_nbr   1.184542
16                race_Asian   1.237443
10          number_inpatient   1.250501
5         num_lab_procedures   1.279866
3        admission_source_id   1.362184
6             num_procedures   1.363202
1          admission_type_id   1.375315
11          number_diagnoses   1.467877
4           time_in_hospital   1.503460
19                race_Other   1.745087
7            num_medications   1.868416
18             race_Hispanic   2.140228
13               diabetesMed   2.604024
12                    change   3.912090
15      race_AfricanAmerican  10.594320
17            race_Caucasian  11.587340